src/prettyprint.rs

```diff
@@ -1,10 +1,398 @@
 use crate::{ast, context};
 use failure::Error;
 
-pub fn prettyprint<'f>(
+use crate::visitor::NodeKind;
+
+struct IndentPrinter<'w> {
+    writer: &'w mut dyn std::io::Write,
+    indent: usize,
+    indent_on_next_write: bool,
+}
+
+impl<'w> IndentPrinter<'w> {
+    fn new(writer: &'w mut dyn std::io::Write) -> IndentPrinter<'w> {
+        IndentPrinter { writer, indent: 0, indent_on_next_write: false }
+    }
+
+    fn print(&mut self, args: std::fmt::Arguments<'_>) {
+        self.indent_if_required();
+        write!(self.writer, "{}", args);
+    }
+
+    fn print_str(&mut self, str: &str) {
+        self.indent_if_required();
+        write!(self.writer, "{}", str);
+    }
+
+    fn newline(&mut self) {
+        writeln!(self.writer);
+        self.indent_on_next_write = true;
+    }
+
+    fn println(&mut self, args: std::fmt::Arguments<'_>) {
+        self.print(args);
+        self.newline();
+    }
+
+    fn indent_if_required(&mut self) {
+        if self.indent_on_next_write {
+            write!(self.writer, "{}", &"\t".repeat(self.indent));
+            self.indent_on_next_write = false;
+        }
+    }
+
+    fn indent(&mut self) { self.indent_var(1); }
+
+    fn outdent(&mut self) { self.indent_var(-1); }
+
+    fn indent_var(&mut self, x: isize) {
+        let mut indent = self.indent as isize;
+        indent += x;
+        if indent < 0 {
+            panic!("setting indent below 0");
+        }
+        self.indent = indent as usize;
+    }
+}
+
+pub fn prettyprint<'f, 'c>(
     program: &ast::Program<'f>,
-    _context: &context::Context<'f>,
+    _context: &context::Context<'c>,
 ) -> Result<(), Error> {
-    println!("way to go:\n{:#?}", program);
+
+    let mut stdout = std::io::stdout();
+    let mut printer = IndentPrinter::new(&mut stdout);
+    do_prettyprint(&NodeKind::from(program), &mut printer);
     Ok(())
 }
+
+#[allow(clippy::cyclomatic_complexity)]
+fn do_prettyprint(n: &NodeKind<'_, '_>, printer: &mut IndentPrinter<'_>) {
```

Comment:
Why not a member function on the printer?

Reply:
The printer is a standalone, reusable component. Why mixing things?

Reply:
It just seems weird having a bunch top-level functions and passing the same object to each and every one of it. Maybe define a `struct PrettyPrinter(IndentPrinter<'_>)` and make the member functions of those. But not today.

Reply:
I am voting for not a member function but its own module. `print::pretty::format` oder something like that.

Reply:
I'll take care of that.

Reply:
done in 9086d03edebe2236acb2db1fb3b21dc251a74e28